### PR TITLE
Update AAD auto-config conditional and split OAuth2AuthorizedClientManager configuration

### PR DESCRIPTION
**Pull request by moarychan**
Fixes https://github.com/Azure/azure-sdk-for-java/issues/26915.

Changelog:
- Update AAD, AAD B2C auto-config condition.
- Split OAuth2AuthorizedClientManager configuration based on application type.

### CODE REVIEW

File: sdk/spring/spring-cloud-azure-autoconfigure/src/test/java/com/azure/spring/cloud/autoconfigure/aad/implementation/conditions/AllWebApplicationConditionTests.java [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+            .withPropertyValues(
+                "spring.cloud.azure.active-directory.credential.client-id = fake-client-id",
+                "spring.cloud.azure.active-directory.application-type=web_application_and_resource_server")
+            .withUserConfiguration(WebApplicationConditionConfig.class)
+            .run(assertConditionMatch(true));
+    }
+
+    @Configuration
+    @Conditional(AllWebApplicationCondition.class)
+    static class WebApplicationConditionConfig extends Config { }
```
Comment (chenrujun) [inline sdk/spring/spring-cloud-azure-autoconfigure/src/test/java/com/azure/spring/cloud/autoconfigure/aad/implementation/conditions/AllWebApplicationConditionTests.java:84]: Keep it align to condition class name. Here is should be `AllWebApplicationConditionConfig`. 
No need to change it if you rename `AllWebApplicationConditionConfig` to `WebApplicationCondition`.

**Review by chenrujun — CHANGES_REQUESTED** [review]

File: sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/aad/AADAuthenticationFilterAutoConfiguration.java [hunk 2]
```diff
@@ -35,7 +35,7 @@
  * spring.cloud.azure.active-directory.session-stateless=true}. Otherwise, {@link AADAuthenticationFilter} will be configured.
  */
 @Configuration(proxyBeanMethods = false)
-@ConditionalOnWebApplication
+@ConditionalOnWebApplication(type = ConditionalOnWebApplication.Type.SERVLET)
```
[the remark this answers is not in the document]
Comment (moarychan) [reply sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/aad/AADAuthenticationFilterAutoConfiguration.java:38]: Yes, forgot this one.

File: sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/aad/implementation/conditions/AllWebApplicationCondition.java [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+
+import java.util.function.Function;
+
+import static com.azure.spring.cloud.autoconfigure.aad.properties.AADApplicationType.RESOURCE_SERVER;
+import static com.azure.spring.cloud.autoconfigure.aad.properties.AADApplicationType.RESOURCE_SERVER_WITH_OBO;
+
+/**
+ * Web application or all in scenario condition.
+ */
+public final class AllWebApplicationCondition extends AbstractApplicationTypeCondition {
```
[the remark this answers is not in the document]
Comment (moarychan) [reply sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/aad/implementation/conditions/AllWebApplicationCondition.java:16]: Good.

File: sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/aad/implementation/conditions/AbstractApplicationTypeCondition.java [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+/**
+ * Abstract class condition for each application type scenario.
+ */
+public abstract class AbstractApplicationTypeCondition extends SpringBootCondition {
+
+    /**
+     * Return function that application type does not match the condition.
+     * @return the no match condition function.
+     */
+    protected abstract Function<AADApplicationType, Boolean> getNoMatchCondition();
```
[the remark this answers is not in the document]
Comment (moarychan) [reply sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/aad/implementation/conditions/AbstractApplicationTypeCondition.java:27]: I use `abstract boolean isNonTargetApplicationType(AADApplicationType applicationType); ` instead.

[the remark this answers is not in the document]
Comment (chenrujun) [reply sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/aad/implementation/conditions/AbstractApplicationTypeCondition.java]: `isTargetApplicationType` is easier to understand when implement this method.
About the hint in Intellij, can we fix by creating another method:
```
abstract boolean isTargetApplicationType(AADApplicationType applicationType);

private boolean isNotTargetApplicationType(AADApplicationType applicationType){ // Non -> Not
    return !isNonTargetApplicationType(applicationType);
}

```

File: sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/aad/configuration/AADOAuth2ClientConfiguration.java [hunk 5]
```diff
@@ -28,60 +31,146 @@
...
-     * @return OAuth2AuthorizedClientRepository bean
+     * Web application scenario, OAuth2AuthorizedClientManager configuration for AAD.
      */
-    @Bean
-    @ConditionalOnMissingBean
-    public OAuth2AuthorizedClientRepository oAuth2AuthorizedClientRepository() {
-        return new JacksonHttpSessionOAuth2AuthorizedClientRepository();
+    @Configuration(proxyBeanMethods = false)
+    @Conditional(WebApplicationWithoutResourceServerCondition.class)
+    public static class WebApplicationOAuth2AuthorizedClientManagerConfiguration {
```
Comment (chenrujun) [inline sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/aad/configuration/AADOAuth2ClientConfiguration.java:72]: nit: Rename to `WebApplicationWithoutResourceServerOAuth2AuthorizedClientManagerConfiguration`

**Review by chenrujun — CHANGES_REQUESTED** [review]

**Review by chenrujun — APPROVED** [review]
LGTM.

**Comment by saragluna** [comment]
@moarychan, will this be targeting both 3.x and 4.x?

**Comment by moarychan** [comment]
As per discussed, this is an enhancement for AAD, only target the 4.0 branch enough.

**Pull request merged**